gms/gossiper.cc

```diff
@@ -421,6 +425,120 @@ future<> gossiper::handle_echo_msg() {
+                {response::leaving_nodes_hash, format("{}", tm.get_leaving_endpoints_hash())},
+                {response::pending_ranges_hash, format("{}", tm.get_pending_ranges_hash())}
+            }};
+            return resp;
+        });
+    }).then([] (std::vector<cluster_ring_status_response> results) {
+        bool same_results = std::adjacent_find(results.begin(), results.end(),
+                std::not_equal_to<cluster_ring_status_response>()) == results.end();
+        results.front().status[response::same_on_all_shards] = same_results ?
+                response::same_on_all_shards_yes : response::same_on_all_shards_no;
+        return results.front();
+    });
+}
+
+// Send CLUSTER_RING_STATUS RPC verb to remote nodes to query the cluster
+// status. If local node has the same cluster ring status as the remote nodes,
+// it means the cluster has the same view on the cluster ring status.
+static future<> wait_for_cluster_ring_status_to_match(gossiper& g) {
+    return seastar::async([&g] {
+        while (true) {
+            auto local = get_cluster_ring_status(g).get0();
+            auto myip = g.get_broadcast_address();
+            if (!local.is_same_on_all_shards()) {
+                logger.debug("Node {}, status = {}, wait for all the shards have the same hashes", myip, local.status);
+                sleep_abortable(std::chrono::milliseconds(100)).get();
+                continue;
+            }
+            auto nodes_to_check = boost::copy_range<std::set<gms::inet_address>>(g.endpoint_state_map | boost::adaptors::map_keys);
+            for (auto& seed : g.get_seeds()) {
+                nodes_to_check.insert(seed);
+            }
+            nodes_to_check.erase(myip);
+            auto node_to_be_replaced = g.get_node_to_be_replaced();
+            if (node_to_be_replaced) {
+                nodes_to_check.erase(*node_to_be_replaced);
+            }
+            size_t nodes_down = 0;
+            std::unordered_map<gms::inet_address, cluster_ring_status_response> queries;
+            parallel_for_each(nodes_to_check, [&nodes_down, &queries] (const auto& node) {
+                return netw::get_local_messaging_service().send_cluster_ring_status(netw::msg_addr(node)).then([node, &nodes_down, &queries] (cluster_ring_status_response result) {
+                    queries[node] = std::move(result);
+                }).handle_exception([node, &nodes_down] (std::exception_ptr ep) {
+                    try {
+                        std::rethrow_exception(ep);
+                    } catch (seastar::rpc::unknown_verb_error& e) {
+                        logger.info("Node {} does not support cluster_ring_status: {}", node, e);
+                        throw e;
+                    } catch (seastar::rpc::closed_error& e) {
+                        nodes_down++;
+                    } catch (...) {
+                        logger.info("Failed to send cluster_ring_status to {}: ", node, std::current_exception());
```

Comment:
Should that be error()?

Reply:
Is this something we can recover? I think yes. We can fallback to the old method.

Reply:
error() log level is not only for unrecoverable errors. It's for when something goes not as it's supposed to.

Reply:
In this case, I think a warn is enough.

Reply:
warn() is enough.